### ISSUE TITLE
Where are KSampler Previews saved? 

### ISSUE THREAD

**Issue by glyfspace** [comfyanonymous/ComfyUI#3400]
I am deploying comfy via a cloud server and trying to save the previews for KSampler during each step so that I can show the image result progress over time. Does anyone know where the KSampler previews are saved or how to access that image data?

**Comment by comfyanonymous** [comment]
They are not saved anywhere, those previews are image information sent with the websocket.

**Comment by ltdrdata** [comment]
> I am deploying comfy via a cloud server and trying to save the previews for KSampler during each step so that I can show the image result progress over time. Does anyone know where the KSampler previews are saved or how to access that image data?

If you want to save progress. You can use KSampler Progress node in Inspire Pack.

**Comment by seghier** [comment]
> image information sent with the websocket.

How to get the final image with python?

**Comment by bigcat88** [comment]
> > image information sent with the websocket.
> 
> How to get the final image with python?

Just open with Pillow the buffer received for message `PREVIEW_IMAGE` or `UNENCODED_PREVIEW_IMAGE` and that is it?

**Comment by seghier** [comment]
Thanks @bigcat88 but i don't get it
How we can use PREVIEW_IMAGE  to get the final image?
in the example file they track the curent node
```
def get_images(ws, prompt):
    prompt_id = queue_prompt(prompt)['prompt_id']
    output_images = {}
    current_node = ""

    while True:
        out = ws.recv()
        if isinstance(out, str):
            message = json.loads(out)
            if message['type'] == 'progress':
                data = message['data']
                current_step = data['value']
                if current_step != 0:
                    print(str(int(100*current_step/steps))+'% Step: ', current_step, ' of: ', data['max'])

            if message['type'] == 'executing':
                data = message['data']
                if data['prompt_id'] == prompt_id:
                    if data['node'] is None:
                        break #Execution is done
                    else:
                        current_node = data['node']
        else:
            if current_node == 'save_image_websocket_node':
                images_output = output_images.get(current_node, [])
                images_output.append(out[8:])
                output_images[current_node] = images_output
        
    return output_images
```

**Comment by bigcat88** [comment]
From the example:

<img width="1106" alt="image" src="https://github.com/comfyanonymous/ComfyUI/assets/13381981/bdaf89ab-0287-41e0-83f0-36edde43c8d5">

**Comment by bigcat88** [comment]
The simple algorithm is like this:

1. if `bytes` arrived via websocket(add check with `isinstance` for that), check the first 4 bytes, if there is a number equal to ONE, then it is **PREVIEW_IMAGE** message.
2. Check the next 4 bytes, if there is a number ONE, it means JPEG, if the number is 2, then it means PNG (this can be seen in my previous message from the ComfyUI source codes)

P.S: the example in ComfyUI simply skips all these 8 bytes: ```out[8:]```

**Comment by glyfspace** [comment]
Is there any way to write the preview images out to the "output" or "temp" folders? Like a custom node or something? that would be the easiest way for accessing the information via a cloud deployment for ComfyUI.

**Comment by ltdrdata** [comment]
> Is there any way to write the preview images out to the "output" or "temp" folders? Like a custom node or something? that would be the easiest way for accessing the information via a cloud deployment for ComfyUI.

Do you want real-time preview information while the progress is being made, or do you simply want to save the process?

If you want real-time feedback, regardless of how it's done, web socket feedback is necessary to fetch information in real-time while the execution is ongoing.

If you simply need images of the process, you can convert intermediate process images obtained through `KSampler Progress` from the previously mentioned Inspire Pack into VAE Decode and save them as `save image`.

**Comment by glyfspace** [comment]
> > Is there any way to write the preview images out to the "output" or "temp" folders? Like a custom node or something? that would be the easiest way for accessing the information via a cloud deployment for ComfyUI.
> 
> Do you want real-time preview information while the progress is being made, or do you simply want to save the process?
> 
> If you want real-time feedback, regardless of how it's done, web socket feedback is necessary to fetch information in real-time while the execution is ongoing.
> 
> If you simply need images of the process, you can convert intermediate process images obtained through `KSampler Progress` from the previously mentioned Inspire Pack into VAE Decode and save them as `save image`.
> 
> ![workflow (14)](https://private-user-images.githubusercontent.com/128333288/328069297-1bdcc262-5a3f-4a09-8514-781da83da384.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1NzE4MTAsIm5iZiI6MTcxNTU3MTUxMCwicGF0aCI6Ii8xMjgzMzMyODgvMzI4MDY5Mjk3LTFiZGNjMjYyLTVhM2YtNGEwOS04NTE0LTc4MWRhODNkYTM4NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxM1QwMzM4MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hODIyOTMxNTQ5ZDE5MmY0Nzk0YTdkMTFkYjg1ZDM1MzJiYjI3M2E3M2RhNzFmMDliZGYxNzU5ODg1ZWEzMWExJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.HzSVw1m7x3XAAEjJ8h9KCYxGlXnx4o7L-kPnDWWTlXs)

This progress latent is saved after the entire generation process right? I can't use this to show the user (I am deploying comfy via cloud) the overall progress of the result doing this right? 

Thanks!

**Comment by glyfspace** [comment]
Yes I am looking for real-time feedback. Did anyone manage to implement that? I am going to take a crack at it this week.

**Comment by tushar221997** [comment]
Was anyone able to get the real time feedback for deployment?

**Comment by pedrocabral** [comment]
> Is there any way to write the preview images out to the "output" or "temp" folders?

Yes.
In the file [latent_preview.py](https://github.com/comfyanonymous/ComfyUI/blob/a3100c8452862e914996648e0fbc56098ab26b60/latent_preview.py#L91-L107), in the `prepare_callback` function, there is a `callback` inner function which every time it is updated (one for step), it decodes the preview image as a Pillow Image object into a tuple (2nd item) that gets assigned to the `preview_bytes` variable. The preview image can then just be saved like this:

`preview_bytes[1].save("your_file_path/image_name.jpeg")`

In this case, with a fixed path/filename, every new preview image that gets saved will overwrite the one saved in the previous step.
To get around this, the callback function has a step counter, which can be used to name each image sequentially.

In the example below, the preview image is saved in the ComfyUI **temp** folder every time it updates, in real time.
For each full image generation run, **all** preview images are saved, one for each step, sequentially. New preview images will overwrite old ones having the same step number.

In the beginning of the file, add a global variable `SAVE_PREVIEWS = True`.
This could be implemented as argument settings to let the preview saving functionality be enabled or disabled from the args.

```
MAX_PREVIEW_RESOLUTION = args.preview_size    <- already in the code, uses arguments to set the max resolution
SAVE_PREVIEWS = True                          <- add this (can also be implemented as argument settings)
```

```python
def prepare_callback(model, steps, x0_output_dict=None):
    preview_format = "JPEG"
    if preview_format not in ["JPEG", "PNG"]:
        preview_format = "JPEG"

    previewer = get_previewer(model.load_device, model.model.latent_format)

    pbar = comfy.utils.ProgressBar(steps)
    def callback(step, x0, x, total_steps):
        if x0_output_dict is not None:
            x0_output_dict["x0"] = x0

        preview_bytes = None
        if previewer:
            preview_bytes = previewer.decode_latent_to_preview_image(preview_format, x0)
            if SAVE_PREVIEWS:
                ext, image, _ = preview_bytes
                image.save(folder_paths.os.path.join(folder_paths.get_temp_directory(), f"preview_{step:0>2}.{ext}".lower()))
        pbar.update_absolute(step + 1, total_steps, preview_bytes)
    return callback
```
